TombEngine/Objects/TR3/Entity/tr3_compies.cpp

```diff
@@ -0,0 +1,330 @@
+	constexpr auto COMPSOGNATHUS_JUMP_ATTACK_CHANCE = 0x1000;
+	constexpr auto COMPSOGNATHUS_ATTACK_CHANCE = 31;
+
+	constexpr auto COMPSOGNATHUS_HIT_FLAG = 1;
+
+	const auto CompyBite = BiteInfo(Vector3(0.0f, 0.0f, 0.0f), 2);
+	const auto CompyAttackJoints = std::vector<unsigned int>{ 1, 2 };
+	
+	enum CompyState
+	{
+		COMPSOGNATHUS_STATE_IDLE = 0,
+		COMPSOGNATHUS_STATE_RUN = 1,
+		COMPSOGNATHUS_STATE_JUMP_ATTACK = 2,
+		COMPSOGNATHUS_STATE_ATTACK = 3,
+	};
+
+	enum  CompyAnim
+	{
+		COMPSOGNATHUS_ANIM_IDLE = 0,
+		COMPSOGNATHUS_ANIM_RUN = 1,
+		COMPSOGNATHUS_ANIM_PREPARE_JUMP_ATTACK = 2,
+		COMPSOGNATHUS_ANIM_JUMP_ATTACK = 3,
+		COMPSOGNATHUS_ANIM_JUMP_ATTACK_LAND = 4,
+		COMPSOGNATHUS_ANIM_ATTACK = 5,
+		COMPSOGNATHUS_ANIM_DIE = 6,
+		COMPSOGNATHUS_ANIM_IDLE_TO_RUN = 7,
+		COMPSOGNATHUS_ANIM_RUN_TO_IDLE = 8,
+	};
+
+	enum CompyTarget
+	{	
+		ATTACK_CADAVER = 0,
+		ATTACK_LARA = 1
+	};
+
+	void InitialiseCompsognathus(short itemNumber)
+	{	
+		auto& item = g_Level.Items[itemNumber];
+		InitialiseCreature(itemNumber);
+
+		//set friendly
+		item.ItemFlags[1] = ATTACK_CADAVER;		
+	}
+
+	void CompsognathusControl(short itemNumber)
+	{
+		if (!CreatureActive(itemNumber))
+			return;
+
+		auto* item = &g_Level.Items[itemNumber];
+		auto* creature = GetCreatureInfo(item);
+		int angle, head, neck, tilt, random, RoomNumber, target;
+		Vector3 cadaverCoordinates;
+		
+		head = neck = angle = tilt = 0;
+
+		if (item->HitPoints <= 0)
+		{
+			if (item->Animation.ActiveState != COMPSOGNATHUS_STATE_IDLE)
+			{
+				SetAnimation(item, COMPSOGNATHUS_ANIM_DIE);
+			}
+		}
+		else
+		{
+			AI_INFO AI;
+			CreatureAIInfo(item, &AI);
+
+			if (AI.ahead)
+				head = AI.angle;
+
+			GetCreatureMood(item, &AI, true);
+			CreatureMood(item, &AI, true);
+
+			if (cadaverCoordinates == Vector3::Zero)
+			{
+				float minDistance = INFINITY;
+
+				int i;
+				ItemInfo* targetItem;
+				for (i = 0, targetItem = &g_Level.Items[0]; i < g_Level.NumItems; i++, targetItem++)
+				{
+					if (targetItem->ObjectNumber == NO_ITEM || targetItem->Index == itemNumber || targetItem->RoomNumber == NO_ROOM)
+						continue;
+
+					if (SameZone(creature, targetItem) )
+					{
+						int distance = Vector3i::Distance(item->Pose.Position, targetItem->Pose.Position);
+
+						if (distance < minDistance && targetItem->Effect.Type == EffectType::Cadaver)
+						{
+							cadaverCoordinates = Vector3(targetItem->Pose.Position.x, targetItem->Pose.Position.y, targetItem->Pose.Position.z);
+							minDistance = distance;
+							item->ItemFlags[1] = ATTACK_CADAVER;							
+						}
+					}
+				}
+			}
+
+			creature->Enemy = LaraItem;
+
+			target = item->ItemFlags[1];
+
+			if (creature->HurtByLara)
+				AlertAllGuards(itemNumber);
+
+			if (creature->Mood == MoodType::Bored && item->ItemFlags[1] == ATTACK_CADAVER)
+			{
+				int dx = cadaverCoordinates.x - item->Pose.Position.x;
+				int dz = cadaverCoordinates.z - item->Pose.Position.z;
+				AI.distance = (dx * dx) + (dz * dz);
```

Comment:
use SQUARE() if you want to multiply a value by itself

Reply:
fixed